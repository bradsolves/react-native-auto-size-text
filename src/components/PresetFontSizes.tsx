import * as React from 'react'
import { NativeSyntheticEvent, Text, TextLayoutEventData } from 'react-native'

import { AutoSizeTextProps } from '../types'

const PresetFontSizes = (props: AutoSizeTextProps) => {
  const { fontSizePresets, children, style, numberOfLines } = props
  const [currentFont, setCurrentFont] = React.useState<number>(
    fontSizePresets![0] as number
  )
  const [currentIndex, setCurrentIndex] = React.useState(0)

  const handleResizing = (e: NativeSyntheticEvent<TextLayoutEventData>) => {
    const { lines } = e.nativeEvent
    if (lines.length > (numberOfLines as number)) {
      if (currentIndex < fontSizePresets!.length - 1) {
        const updatedIndex = currentIndex + 1
        setCurrentIndex(updatedIndex)
        setCurrentFont(fontSizePresets![updatedIndex])
      }
    }
  }

  return (
    <Text
      numberOfLines={numberOfLines}
      style={[
        style,
        {
          fontSize: currentFont,
        },
      ]}
      onTextLayout={handleResizing}
    >
      {children}
    </Text>
  )
}

export default PresetFontSizes
