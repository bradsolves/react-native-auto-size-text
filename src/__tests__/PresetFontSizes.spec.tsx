import { cleanup, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import PresetFontSizes from '../components/PresetFontSizes';

afterEach(cleanup);

describe('PresetFontSizes snapshot', () => {
  it('PresetFontSizes renders correctly', () => {
    const tree = renderer
      .create(
        <PresetFontSizes fontSizePresets={[50, 40, 30]} numberOfLines={3}>
          Lorem impsum
        </PresetFontSizes>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('PresetFontSizes mode', () => {
  it('should pass the right props to Text component inside PresetFontSizes', async () => {
    const mockContent = 'Lorem ipsum';
    const mockNumberOfLines = 3;
    const fontSizePresets = [50, 42, 14];
    const mockColor = 'red';

    const { getByTestId } = render(
      <PresetFontSizes
        style={{ color: mockColor }}
        fontSizePresets={fontSizePresets}
        numberOfLines={mockNumberOfLines}
      >
        {mockContent}
      </PresetFontSizes>
    );

    const textComponent = getByTestId('preset-font-sizes');
    const content = textComponent.props.children;
    const numberOfLines = textComponent.props.numberOfLines;
    const adjustsFontSizeToFit = textComponent.props.adjustsFontSizeToFit;
    const fontSize = textComponent.props.style[1].fontSize;
    const color = textComponent.props.style[0].color;

    await waitFor(() => {
      expect(content).toBe(mockContent);
      expect(numberOfLines).toBeFalsy();
      expect(fontSize).toBe(50);
      expect(color).toBe(mockColor);
      expect(adjustsFontSizeToFit).toBeFalsy();
    });
  });
});
