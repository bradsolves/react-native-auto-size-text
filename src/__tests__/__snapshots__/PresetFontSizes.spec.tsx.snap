// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PresetFontSizes snapshot PresetFontSizes renders correctly 1`] = `
<Text
  onTextLayout={[Function]}
  style={
    Array [
      undefined,
      Object {
        "fontSize": 50,
      },
    ]
  }
  testID="preset-font-sizes"
>
  Lorem impsum
</Text>
`;
